File: Python/59.Spiral Matrix II.py
#数字按顺时针的顺序填满一个正方形
class Solution(object):
    def generateMatrix(self, n):
        matrix = [[0]*n for i in range(n)]
        i , j , x ,c = 0 , 0 , 1 , 0
        while x<=n*n:
            while x<=n*n and j<n-c:
                matrix[i][j] = x
                j , x = j+1 , x+1
            j , i =  j-1 , i+1
            while x<=n*n and i<n-c:
                matrix[i][j] = x
                i , x = i+1 , x+1
            i , j=  i-1 , j-1
            while x<=n*n and j>=c:
                matrix[i][j] = x
                j , x = j-1 , x+1
            j ,i = j+1, i-1
            while x<=n*n and i>c:
                matrix[i][j] = x
                i , x = i-1 , x+1
            i ,j = i+1 , j+1
            c += 1
        return matrix

s = Solution()
print(s.generateMatrix(4))

File: C++/Array/59.Spiral Matrix II.cpp
/*
给出数字n，返回螺旋状的二维数组，元素从1到n^2。
*/

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	vector<vector<int>> generateMatrix(int n) {
		if(n==0)    {vector<vector<int>> result;    return result;}
		vector<vector<int>> result(n,vector<int>(n));
        int m=--n,rowstart=0,colstart=0,num=1;
		while (rowstart<=m && colstart<=n)
		{
			for (int j = colstart; j <= n; ++j)
			{
				result[rowstart][j] = num++;
			}
			++rowstart;
			for (int i = rowstart; i <= m; ++i)
			{
				result[i][n] = num++;
			}
			--n;
			if(rowstart<=m)
			{
				for (int j = n; j >= colstart; --j)
				{
					result[m][j] = num++;
				}
			
			}
			--m;
			if(colstart<=n)
			{
				for (int i = m;i >= rowstart; --i)
				{
					result[i][colstart] = num++;
				}
			}
			++colstart;
		}
		return result;  
    }
};

int main()
{
	Solution s;
	vector<vector<int>> out= s.generateMatrix(5);
	for(int i=0; i < out.size(); ++i)
	{
		for(int j=0; j < out[i].size(); ++j)
		{
			cout<<out[i][j]<<"\t";
		}
		cout<<endl;
	}
}

